### ISSUE TITLE
Need help in linking with HA

### ISSUE THREAD

**Issue by stauffenberg2020** [MattJeanes/TeslaMateAgile#63]
Hi All,

I am trying to get prices through [hass-monta](https://github.com/nickknissen/hass-monta) and I only have a "Monta - Personal Wallet" sensor/entity. The output from that monta API call is like some YAML format as mentioned in https://github.com/nickknissen/hass-monta/issues/65#issuecomment-2115166644

I am trying to link these prices to TeslaMateAgile and I am wondering how it was done [previously](https://github.com/MattJeanes/TeslaMateAgile/issues/56#issuecomment-2200362072).

I also see that @tougher has been working on a [fork ](https://github.com/MattJeanes/TeslaMateAgile/issues/47#issuecomment-1873923542)to support the data from Monta into TeslaMateAgile - But I am not sure if this has been completed.

Can @MattJeanes or @tougher or @goncal you help me a bit here?

Thanks in advance

**Comment by MattJeanes** [comment]
Hey, I don't know much about Monta or hass-monta but it sounds like what you probably need to do is create a Home Assistant template entity that references the appropriate field from the big API response entity, it should be possible to do. TeslaMateAgile requires a number / input_number type entity to work from so if you can extract it out into one of those it should work.

**Comment by MattJeanes** [comment]
Okay so pretty much what it's looking for is an entity (with recording turned on so it can see historically too, it's on by default unless you turn it off in HA configuration.yaml btw) that captures the current price of electricity. The Home Assistant history is then queried with the times of your charging session to get the price of electricity at those times and then calculates based on the energy being put into the car how much it cost.

For example, here's what my one looks like that I have set up based off the Octopus Intelligent integration:

![image](https://github.com/user-attachments/assets/9118ef07-4440-4731-80d3-621af23b2c7a)

Here's some documentation on template sensors which may help: https://www.home-assistant.io/integrations/template/

**Comment by MattJeanes** [comment]
Ah I see, TeslaMateAgile isn't really designed to work like that as all the various price providers do is give point in time data for the electricity price as the purpose of it is to calculate the charge cost from that data.

If you already have the charge cost then you'll probably need something else to do that, as all that ultimately needs to happen is edit the `cost` column in the `charging_processes` table in the database with the cost number. It could potentially be implemented in a code change to TeslaMateAgile to allow a provider to supply the entire cost I suppose.

How might you correlate the charge in TeslaMate with Monta, is there a reliable method of doing that? We could try and guess using the time or something but might not be super reliable.

**Comment by MattJeanes** [comment]
Yeah that's the main concern really, depends how close the times are from the Monta data to the TeslaMate tracked data for the same charge. If it's close enough we could probably correlate them automatically, and just ignore ones we cannot find a good match for. Looking at that output example you gave probably could yeah.

Supercharging would be the same, but we'd need to access the Tesla charging history which is a little more complicated now since Tesla locked down their APIs and require people to register applications. It's doable though, if you already have e.g. Tesla connected via Fleet API in Home Assistant then you could use that same thing to get charging history. More info here https://developer.tesla.com/docs/fleet-api/getting-started/what-is-fleet-api

It's certainly an interesting feature request and supercharging especially I'm sure people would love to see!

Since I don't have Monta, would you mind exporting an example response from yours and then also export a couple of records from the charging_process table to me? Ideally as a CSV but anything will do, and I'll see what I can do with it!

**Comment by MattJeanes** [comment]
That's great thank you, if you could also export some data from the TeslaMate charging_process table (there's a guide on fixing data in the readme for how to access the DB) that matches up with that data from Monta I'll see how easy it'll be to implement. 

I don't need the client id/secret for now but I'll let you know if that changes. I'm hoping they have some kind of test or demo token I can use in the mean time but I haven't really looked at the API yet.

I'll implement it directly with calls to Monta, going via Home Assistant would actually probably make it harder to do.

Also the text file you sent is written in the YAML format for your reference 🙂

**Comment by MattJeanes** [comment]
Fantastic, that's perfect thank you. I'll let you know when I have something, but just to set expectations this is very much a spare time project - and spare time is something of a scarce resource to me 🙂

That said I do hope to look at this soon, just can't promise anything right now!

**Comment by MattJeanes** [comment]
Hey, I've just tried to have a look at their API guide but I can't seem to find the option to make an application in their UI for testing.

If you're able to make one (and can restrict the application and/or trust me enough) then you could email over a client id/secret for an application and I can use that for testing. If you are happy to do that my email is on my GitHub profile and you can always remove / revoke the application afterwards.

This is what I see in the UI:

![image](https://github.com/user-attachments/assets/3a625430-ed5f-455c-bf07-1603d088d491)

But apparently I should see this:

![image](https://github.com/user-attachments/assets/b8458809-5131-4895-9c47-c038dbddad42)

🤷🏻

**Comment by MattJeanes** [comment]
Thank you, that helps a lot! I'll look into this when a get a spare moment, hopefully tonight 🙂

**Comment by MattJeanes** [comment]
Hey @stauffenberg2020, I've had some time to look at this now and I've implemented it in release candidate [`v1.16.0-rc.1`](https://github.com/MattJeanes/TeslaMateAgile/releases/tag/v1.16.0-rc.1)

Can you please give it a go and let me know how it goes! Check out the README for the updated documentation on how to use it. Please also note the new `TeslaMate__MatchingToleranceMinutes` setting which is set to 30 minutes by default, I am not sure if that'll be enough or not.

Any questions or problems please let me know!

**Comment by tougher** [comment]
Nice, I will test as well.
I already did a similar implementation, but it didn't really work because I was missing the new `IWholePriceDataService` and I didn't dare to make bigger breaking changes :)

**Comment by MattJeanes** [comment]
@stauffenberg2020 interesting, I've put your dates plus your 120 minute tolerance into my integration test (which is using your client id / secret at the moment) and it is picking up a charge for you:

![image](https://github.com/user-attachments/assets/3de0ea66-89a8-49e7-bee1-4fc657e18e8d)

It seems that the end date of that charge is 11:21 UTC - whereas the tolerance would only allow up to 10:57 UTC (8:57 + 120 minutes) which is the problem here.

Here's the full details of that charge from the Monta API (some details redacted!)

```json
{
  "id": xxxxxxx,
  "chargePointId": xxxxx,
  "createdAt": "2024-10-17T05:51:27Z",
  "updatedAt": "2024-10-17T11:22:03Z",
  "cablePluggedInAt": "2024-10-17T05:52:03Z",
  "startedAt": "2024-10-17T05:51:27Z",
  "stoppedAt": "2024-10-17T11:21:56Z",
  "fullyChargedAt": null,
  "failedAt": null,
  "timeoutAt": null,
  "state": "completed",
  "consumedKwh": 35.66,
  "kwhLimit": null,
  "startMeterKwh": 0,
  "endMeterKwh": 0,
  "cost": 10.61,
  "price": 26.82,
  "priceLimit": 40,
  "averagePricePerKwh": 0.75,
  "averageCo2PerKwh": 44.54,
  "averageRenewablePerKwh": 90.72,
  "failureReason": null,
  "stopReason": "api",
  "note": null,
  "currency": {
    "identifier": "dkk",
    "name": "Danish kroner",
    "decimals": 2
  },
  "soc": {
    "percentage": 89,
    "source": "vehicle"
  },
  "socLimit": null
}
```

An interesting thing I've just noticed is that there is a cost field and a price field which have two different values - I'm not sure which one is correct, right now we are using the `cost` field.

Maybe we need to have separate tolerances for the start and end? The start seems very close (within 30 seconds!) and that's probably the more important one. I should probably also be filtering for completed charges which I'm not currently doing and there's probably a bit better logging I could do also 😄 

We could also potentially try and match filter based on the `consumedKwh` too, that could be interesting for providers that support it.

What do you think?

Thanks @tougher also for having a look!

**Comment by MattJeanes** [comment]
@stauffenberg2020 I've just made some changes I hope will fix the issues 😄 

Check out [`v1.16.0-rc.2`](https://github.com/MattJeanes/TeslaMateAgile/releases/tag/v1.16.0-rc.2) and see if that works for you!

- Added start time / energy matching to replace start time / end time matching when energy data is provided
- Split `TeslaMate__MatchingToleranceMinutes` into `TeslaMate__MatchingStartToleranceMinutes` and `TeslaMate__MatchingEndToleranceMinutes` with new default values of `30` and `120` respectively
- Added setting `TeslaMate__MatchingEnergyToleranceRatio` with default value of `0.1`
- Improved logging for whole cost providers
- Fixed Monta provider looking at the wrong cost field
- Monta provider now only searches for completed charges

**Comment by MattJeanes** [comment]
Awesome!

Yes, the geofence ID is still valid as that is what TeslaMate uses to differentiate different charging locations. Presumably you are using Monta just at home so the home geofence ID is what you want to use.

There is a separate `Monta__ChargePointId` setting which you can optionally set although it seems in your case your charges are all on different ones and there aren't any registered to your account - public charging maybe rather than a home charger?

You can feel free to revoke the client id / secret you sent me now, although it is useful for debugging if I need to make any further tweaks to the Monta provider.

I'll wait to hear from @tougher to see how they are doing and if they see any issues with it, otherwise I'll push it to full release.

Thank you for your patience, sorry it took a while!

**Comment by tougher** [comment]
Really nice work @MattJeanes !

I will continue the testing tonight, but I just noticed that my implementation does set some default header:
```csharp
_client.DefaultRequestHeaders.Add("operator", "monta");
_client.DefaultRequestHeaders.Add("meta", "ios;production;2.11.0;iOS16.2.1;iPhone14,2");
_client.DefaultRequestHeaders.Add("uuid", _settings.Uuid);
_client.DefaultRequestHeaders.Add("timezone", _options.Timezone);
_client.DefaultRequestHeaders.Add("accept-language", "en");
```
Maybe the `timezone` actually does affect date times to and from the API - I hope not.

**Comment by MattJeanes** [comment]
That shouldn't be an issue hopefully, API documentation doesn't say anything about setting headers (other than content-type and authentication).

Times are generally encoded with the timezone information embedded in them so that should also be okay 😄 

Let me know how testing goes! Cheers

**Comment by MattJeanes** [comment]
Ah I see, yes that makes sense, 24 hours should be a big enough window I would imagine for planned charges, it seems unlikely that it would wait more than a day to charge after being plugged in.

We already filter down the most sensible charge from available data so as far as I'm concerned all we'd need to do is change the search parameters and hardcode them to 24 hours before and after.

If you want to throw a PR in feel free, it's only a couple of lines though 😁

**Comment by tougher** [comment]
Sorry for the delay, I had to spend my energy in "rescuing" my vulnerable Home Assistant setup.

I made a PR: https://github.com/MattJeanes/TeslaMateAgile/pull/66
I had more changes, but it turned out that the -24/24 hours change was the only thing necessary.

**Comment by MattJeanes** [comment]
@tougher since you've already tested these changes yourself, now they're merged - do you reckon we're good for a full release now?

**Comment by MattJeanes** [comment]
Hey @tougher sorry to ping, did you think we were all good to release these changes properly now to everyone instead of a pre-release?

**Comment by tougher** [comment]
@MattJeanes yes, I think so. I was actually standby to update [hassio-addon-TeslaMateAgile](https://github.com/tougher/hassio-addon-TeslaMateAgile). Good you pinged me, I missed that your message from last week was a question - sorry.

**Comment by MattJeanes** [comment]
Ah, no worries! That's now released as [v1.16.0](https://github.com/MattJeanes/TeslaMateAgile/releases/tag/v1.16.0)

Thanks for your help both! Appreciate it a lot 🙂
